pkg/operator/apis/monitoring/v1alpha1/types.go

```diff
@@ -333,16 +313,58 @@ func (pm *PodMonitoring) endpontScrapeConfig(index int) (*promconfig.ScrapeConfi
 		Replacement: pm.Name,
 		TargetLabel: "job",
 	})
-	// The instance label being the pod name would be ideal UX-wise. But we cannot be certain
-	// that multiple metrics endpoints on a pod don't expose metrics with the same name. Thus
-	// we have to disambiguate along the port as well.
-	relabelCfgs = append(relabelCfgs, &relabel.Config{
-		Action:       relabel.Replace,
-		SourceLabels: prommodel.LabelNames{"__meta_kubernetes_pod_name", portLabel},
-		Regex:        relabel.MustNewRegexp("(.+);(.+)"),
-		Replacement:  "$1:$2",
-		TargetLabel:  "instance",
-	})
+
+	// Filter targets by the configured port.
+	//
+	// If a port number is configured we don't filter by the __meta_kubernetes_pod_container_port_number
+	// label but instead hardcode it into the config and rely on Prometheus' target deduplication to
+	// ensure only one final target is produced.
+	// We do this because a user may not be able to ensure that all Pods sufficiently declare ports
+	// but should still be able to enforce a specific port being scraped.
+	//
+	// This approach works in our case for two reasons:
+	//   1. if a Pod specifies no ports at all as Prometheus will still generate a single substitute target
+	//   2. we generate a scrape config for each specified endpoint and can thus safely force all candidates
+	//      to have a static port number without preventing other ports from being scraped.
+	if ep.Port.StrVal != "" {
+		portValue, err := relabel.NewRegexp(ep.Port.StrVal)
+		if err != nil {
+			return nil, errors.Wrapf(err, "invalid port name %q", ep.Port)
+		}
+		relabelCfgs = append(relabelCfgs, &relabel.Config{
+			Action:       relabel.Keep,
+			SourceLabels: prommodel.LabelNames{"__meta_kubernetes_pod_container_port_name"},
+			Regex:        portValue,
+		})
+		// The instance label being the pod name would be ideal UX-wise. But we cannot be certain
+		// that multiple metrics endpoints on a pod don't expose metrics with the same name. Thus
+		// we have to disambiguate along the port as well.
+		relabelCfgs = append(relabelCfgs, &relabel.Config{
+			Action:       relabel.Replace,
+			SourceLabels: prommodel.LabelNames{"__meta_kubernetes_pod_name", "__meta_kubernetes_pod_container_port_name"},
+			Regex:        relabel.MustNewRegexp("(.+);(.+)"),
+			Replacement:  "$1:$2",
+			TargetLabel:  "instance",
+		})
+	} else if ep.Port.IntVal != 0 {
+		relabelCfgs = append(relabelCfgs, &relabel.Config{
+			Action:       relabel.Replace,
+			SourceLabels: prommodel.LabelNames{"__meta_kubernetes_pod_name"},
+			Replacement:  fmt.Sprintf("$1:%d", ep.Port.IntVal),
+			TargetLabel:  "instance",
+		})
+		// If the input target was produced by a port declared in the Pod, this port number
```

Comment:
What happens if you don't do this step?

Reply:
It uses the pre-populated __address__ which is whatever port this target was originally generated for.
Say the pod has 2 ports declared, 2 targets would be generated. We hard-override both to the same labels via the config here and Prometheus will dedupe them into a single target.